### PR TITLE
Assertions

### PR DESCRIPTION
**Pull request by danfuzz**
This adds an optional `message` argument to all of `typ`'s assertion
functions.

Hello mike, sho, 

Please review the following commits I made in branch 'danfuzz-assert'.

a810e4a6cbb3fca76eccfb14592da6747581f925 (77 seconds ago)
Add tests.

c478d5f1b099ae1a969e7a2300d227d3a98c43b6 (2 minutes ago)
Finish implementing.

93f3a8ffaa16c07eb71dc07b31716a7829ae82bd (10 minutes ago)
Implement it.

fc589b5e4981d50a898513b1197766a4d5ba1176 (16 minutes ago)
Document what's upcoming.

R=mike
R=sho

### CODE REVIEW

File: lib/typ.js [hunk 1]
```diff
@@ -148,33 +148,36 @@ function extendedTypeOf(value) {
...
-    case REGEXP:
-    case UINT: {
-      message += " (" + value + ")";
-      break;
-    }
-    case ERROR: {
-      if (value.message) {
-        message += " (" + value.message + ")";
+  if (isUndefined(message)) {
+    message = "Expected " + expectedTypeName + "; got " + gotType;
```
Comment (skuwamoto) [inline lib/typ.js]: I was mildly surprised that you did it this way. I think 'expected foo got bar' is still useful even if you have a message passed in.

E.g.,

'Problem with tablename: expected string, got function' or whatever.

Comment (danfuzz) [inline lib/typ.js]: I figured there are probably cases where the original message is unhelpful (e.g. if the value being checked wasn't directly under client control). I think the most straightforward way to get the best of both worlds is to do `%s` substitution in the message, a la `console.log()`.

**Comment by mikefleming** [comment]
LGTM

Whoo thanks!

File: lib/typ.js (hunk 1 above)
Comment (danfuzz) [inline lib/typ.js]: Tweaked! Let me know what you think.

**Comment by mikefleming** [comment]
LGTM
Fancy!

File: lib/typ.js (hunk 1 above)
Comment (skuwamoto) [inline lib/typ.js]: Thanks, dan!!

**Pull request merged**